### PR TITLE
Fix #385: AttributeError on pickling Django models

### LINKED ISSUE #306: Feature proposal: custom serializers
Hey guys,

I have picked up django-cacheops to cache Pandas dataframes stored in my database and it works great!

However, I can't override the default behaviour of pickling the series in Redis. I'd prefer to use an alternative serializer, the msgpack one.

On quick first look it seems fairly simple to expose the serializer method either as a global option or as a parameter to the `cached_as` decorator.

What do you think? Would you be interested in a contribution for this?

### LINKED ISSUE #385: AttributeError: Can't pickle local object 'curry.<locals>._curried'
We got this error when using `curried` method in django model.

For reproduce:
```Python
class Client(models.Model):
    def __init__(self, *args, **kwargs):
        # copied from Django 2.1.5 (not exists in Django 3.1.5 installed by current requirements)
        def curry(_curried_func, *args, **kwargs):
            def _curried(*moreargs, **morekwargs):
                return _curried_func(*args, *moreargs, **{**kwargs, **morekwargs})

            return _curried

        super().__init__(*args, **kwargs)
        setattr(self, '_get_private_data', curry(sum, [1, 2, 3, 4]))
```

Two line test:
```Python
Client.objects.create(name='Client Name')
Client.objects.filter(name='Client Name').cache().first()
```

Test result:
```Bash
  File "/path/to/lib/site-packages/cacheops/query.py", line 58, in cache_thing
    pickle.dumps(data, -1),
AttributeError: Can't pickle local object 'curry.<locals>._curried'
```

Right now I finishing works by fix this and will create PR.
Please waiting any updates inside this issue.

### PR DESCRIPTION
**Pull request by itcrab**
Fixes #385. Closes #306. Helps with #362.

### CODE REVIEW

**Comment by Suor** [comment]
If you are going to implement custom serializer support that should be done in context of #306 and #362. Some principles I look for:
- no new dependencies
- simply interface, preferably a single setting
- no custom serializers embedded into cacheops code (at least for now)

More in line comments.

File: README.rst [hunk 1]
```diff
@@ -76,6 +76,9 @@ Setup redis connection and enable caching for desired models:
     # should be compatible or subclass cacheops.redis.CacheopsRedis
     CACHEOPS_CLIENT_CLASS = 'your.redis.ClientClass'
 
+    # You can use two pickling libs: pickle or dill
+    CACHEOPS_PICKLE_LIB = 'pickle'
```
Comment (Suor) [inline README.rst:80]: I would call it `CACHEOPS_SERIALIZER` and make it support both string and object:

```python
CACHEOPS_SERIALIZER = "pickle"       # default
CACHEOPS_SERIALIZER = "dill"         # use dill
CACHEOPS_SERIALIZER = "your.module"  # use any module with .loads() and .dumps()


# Custom implementation, small and promitive enough to inline in settings
import pickle

class Serializer:
    dumps = lambda obj: pickle.dumps(obj, 3)  # Use non-default protocol
    loads = pickle.loads

CACHEOPS_SERIALIZER = Serializer  
```

Comment (Suor) [reply README.rst:80]: Also, give it a separate section in README.

File: cacheops/query.py [hunk 2]
```diff
@@ -32,6 +31,13 @@
 from .transaction import transaction_states
 from .signals import cache_read
 
+if settings.CACHEOPS_PICKLE_LIB == 'pickle':
+    import pickle
+elif settings.CACHEOPS_PICKLE_LIB == 'dill':
+    import dill as pickle
+else:
+    raise ValueError('settings.CACHEOPS_PICKLE_LIB invalid value (use pickle or dill)')
```
Comment (Suor) [inline cacheops/query.py:39]: Use `import_string()` and do it once somewhere, maybe in `conf`. Using those at module load time will make it harder to test.

See how `CACHEOPS_PREFIX` is handled.

File: setup.py [hunk 3]
```diff
@@ -31,6 +31,7 @@
         'redis>=3.0.0',
         'funcy>=1.8,<2.0',
         'six>=1.4.0',
+        'dill>=0.3.3',
```
Comment (Suor) [inline setup.py:34]: No new dependencies. Most people don't need it.

File: tests/tests.py [hunk 4]
```diff
@@ -638,6 +638,26 @@ def test_365(self):
...
+            Client.objects.filter(name='Client Name').cache().first()
+        self.assertEqual(str(e.exception),
+                         "Can't pickle local object 'Client.__init__.<locals>.curry.<locals>._curried'")
+
+        invalidate_model(Client)
+
+        # case with settings.CACHEOPS_PICKLE_LIB == 'dill'
+        import dill
+        with mock.patch('pickle.loads', new=dill.loads), \
+             mock.patch('pickle.dumps', new=dill.dumps):
```
Comment (Suor) [inline tests/tests.py:656]: If you implement settings more carefully you can use `override_settings()` here.

**Review by Suor — APPROVED** [review]

**Review by Suor — CHANGES_REQUESTED** [review]
See all the things above. Approved by mistake.

File: setup.py (hunk 3 above)
Comment (itcrab) [reply setup.py:34]: How you see picklng not supported by `pickle` lib curried methods without any new dependency? Maybe I don't see right way for it,

Comment (itcrab) [reply setup.py:34]: I see above - doing it in few days. Thank's!

**Comment by itcrab** [comment]
@Suor I finished works by your review - please see my updates.

One comment for your review:
>no new dependencies
>no custom serializers embedded into cacheops code (at least for now)

I remove `dill` dependency by `cacheops` as Python package but inside tests env leave `dill` dependancy and create litle serializer for `dill` lib - it give us opportunity see that issue problem is fixed right now by this PR.

File: README.rst [hunk 5]
```diff
@@ -76,6 +76,10 @@ Setup redis connection and enable caching for desired models:
     # should be compatible or subclass cacheops.redis.CacheopsRedis
     CACHEOPS_CLIENT_CLASS = 'your.redis.ClientClass'
 
+    # To use your own serializer class,
+    # should be compatible or subclass cacheops.serializers.PickleSerializer
+    CACHEOPS_SERIALIZER = 'your.serializers.ClientClass'
```
Comment (Suor) [inline README.rst:81]: It's much easier to simply point to a module. And default will be simply `pickle`.

File: cacheops/conf.py [hunk 6]
```diff
@@ -32,7 +33,7 @@ class Defaults:
 class Settings(object):
     def __getattr__(self, name):
         res = getattr(base_settings, name, getattr(Defaults, name))
-        if name == 'CACHEOPS_PREFIX':
+        if name in ['CACHEOPS_PREFIX', 'CACHEOPS_SERIALIZER']:
             res = res if callable(res) else import_string(res)
```
Comment (Suor) [inline cacheops/conf.py:37]: Resolve if string, leave as is in other case. Will allow to use a module, a class or an instance.

File: cacheops/query.py [hunk 7]
```diff
@@ -52,7 +51,7 @@ def cache_thing(prefix, cache_key, data, cond_dnfs, timeout, dbs=(), precall_key
     load_script('cache_thing', settings.CACHEOPS_LRU)(
         keys=[prefix, cache_key, precall_key],
         args=[
-            pickle.dumps(data, -1),
+            settings.CACHEOPS_SERIALIZER.dumps(data, -1),
```
Comment (Suor) [inline cacheops/query.py:54]: That -1 looks not generic enough

File: README.rst [hunk 8]
```diff
@@ -653,6 +657,46 @@ A ``query`` object passed to callback also enables reflection on used databases
 **NOTE:** prefix is not used in simple and file cache. This might change in future cacheops.
 
 
+Pickle serializer
```
Comment (Suor) [inline README.rst:660]: Might be something else not only pickle.

File: cacheops/simple.py [hunk 9]
```diff
@@ -86,11 +85,11 @@ def get(self, cache_key):
         data = self.conn.get(cache_key)
         if data is None:
             raise CacheMiss
-        return pickle.loads(data)
+        return settings.CACHEOPS_SERIALIZER.loads(data)
 
     @handle_connection_failure
     def set(self, cache_key, data, timeout=None):
-        pickled_data = pickle.dumps(data, -1)
+        pickled_data = settings.CACHEOPS_SERIALIZER.dumps(data, -1)
```
Comment (Suor) [inline cacheops/simple.py:92]: Same as above.

File: cacheops/simple.py [hunk 10]
```diff
@@ -149,7 +148,8 @@ def set(self, key, data, timeout=None):
             # Use open with exclusive rights to prevent data corruption
             f = os.open(filename, os.O_EXCL | os.O_WRONLY | os.O_CREAT)
             try:
-                os.write(f, pickle.dumps(data, pickle.HIGHEST_PROTOCOL))
+                os.write(f, settings.CACHEOPS_SERIALIZER.dumps(
+                    data, settings.CACHEOPS_SERIALIZER.HIGHEST_PROTOCOL))
```
Comment (Suor) [inline cacheops/simple.py:152]: Also weird to require `HIGHEST_PROTOCOL` from any serializer.

**Review by Suor — CHANGES_REQUESTED** [review]

**Pull request merged**